src/interpreter/mod.rs

```diff
@@ -822,14 +802,14 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
         }
     }
 
-    fn eval_lvalue(&mut self, lvalue: &mir::Lvalue<'tcx>) -> EvalResult<'tcx, Lvalue> {
+    fn eval_lvalue(&mut self, lvalue: &mir::Lvalue<'tcx>) -> EvalResult<'tcx, Value> {
```

Comment:
I am really unsure about these changes to `eval_lvalue`. They seem to mix up the distinction between `Value::ByRef(ptr)` and `Value::ByVal(PrimVal::Ptr(ptr))`, and I think we need to have a consistent interpretation of `Value` throughout the codebase.

Say you had a `Value` that represents an `i32`. This could be represented either by `Value::ByRef(ptr)` where `ptr` points to somewhere in an allocation that the integer lives, or by `Value::ByVal(PrimVal::I32(n))`. Since these two cases are equivalent, the `Value::ByVal(PrimVal::Ptr(ptr))` can't be equivalent to the first case.

In fact, since `Value::ByVal` doesn't represent a location in memory, an lvalue can never be represented by `Value::ByVal`. But this same argument holds for `Value::ByValPair` - it represents two immediates not living in memory, rather than referring to any memory region.

In general, my current interpretation of `Value` doesn't permit it to directly refer to a dynamically sized region of memory, but lvalues have to.